### ISSUE TITLE
Asynchronous IO

### ISSUE THREAD

**Issue by Mrkol** [assimp/assimp#4255]
As far as I can tell, assimp still does not support truly asynchronous IO. This was discussed in #227 a long time ago, but the feature request was closed for no apparent reason.

My preferred solution to this would be a message passing style interface with callbacks:

- User requests assimp to load a mesh file (no actual work gets done at this point of time)
- Assimp asks the user to get a chunk of data from a certain file at a specified offset (may happen several times)
- User notifies assimp that data is available
- Assimp asks the user to run some heavy-weight function that actually parses data (which the user can run on a separate threadpool, may happen several times as well)
- Assimp notifies the user that the model is ready

An alternative to the fourth step would be for assimp to do it's heavy work inside of the third step, but that would use the user's IO context computational resources and would not be scalable.
Although I have close to zero experience with assimp, so no one should be listening to me.
Opinions and discussion is welcome, but please don't sweep this issue under the rug. We can do a lot better in modern OSes than simply running assimp stuff on a separate thread.

What this feature request is NOT about: any kind of concurrency support from assimp's side. No overengineered built-in threadpools, no automatic load job balancing, nothing. This request is only about "hooks" for integrating either with native system async IO, or preferably a crossplatform library (e.g. latest versions of ASIO).

**Comment by tellypresence** [comment]
@Mrkol Is this still a problem?

**Comment by Mrkol** [comment]
> @Mrkol Is this still a problem?

What do you mean? I haven't kept up with assimp updates in summer time, did this functionality get exposed in some way?

**Comment by turol** [comment]
No, this has not been done and probably will not be since it would require extreme refactoring. However you can fake it with either threads or green threads. See https://www.youtube.com/watch?v=HIVBhKj7gQU
